### PR TITLE
[Fixes #4876] ContentResult forcing chunked encoding

### LINKED ISSUE #4876: ContentResult forcing chunked encoding
It looks like the ContentResult is disabling response buffering in order to skip the memory allocation. But it's not setting the content length before doing the write, so it's forcing the result to be chunked. Would be much better to encode to buffer pooled bytes and set the content length before doing the write from bytes.

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs#L71

Returning an ObjectResult of the string as a workaround gives the desired effect - the response buffering middleware `return "OK"` comes back with Content-Length: 2

### PR DESCRIPTION
**Pull request by kichalla**
@Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 1]
```diff
@@ -67,13 +67,17 @@ public override Task ExecuteResultAsync(ActionContext context)
...
-                return response.WriteAsync(Content, resolvedContentTypeEncoding);
-            }
+                response.ContentLength = data.Length;
 
-            return TaskCache.CompletedTask;
+                return response.Body.WriteAsync(data, 0, data.Length);
+            }
+            else
+            {
+                response.ContentLength = 0;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: This is optional, the servers will do it for you (assuming nobody calls Flush)

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 2]
```diff
@@ -67,13 +67,17 @@ public override Task ExecuteResultAsync(ActionContext context)
 
             if (Content != null)
             {
-                var bufferingFeature = response.HttpContext.Features.Get<IHttpBufferingFeature>();
-                bufferingFeature?.DisableResponseBuffering();
+                var data = resolvedContentTypeEncoding.GetBytes(Content);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: @lodejard asked for pooled byte buffers here. @rynowak ?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: We need to figure out how that works, we can't just do unbounded buffering or unbounded allocation (like this does). 

For now let's just get rid of the call to `DisableResponseBuffering`

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: @Tratcher set me straight on this. Since `ContentResult` only uses strings we can compute the byte count up front, and then using a 'rolling' buffer to iterate through the string. So lets do that.

Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: Sure, sounds good.

**Comment by kichalla** [comment]
Thanks @rynowak ...great points...updating to use alternative approaches...need to figure out about BOM...

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 3]
```diff
@@ -69,60 +68,30 @@ public class ContentResult : ActionResult
...
-                        var numOfCharsToCopy = Math.Min((Content.Length - sourceIndex), charBuffer.Length);
-
-                        Content.CopyTo(
-                            sourceIndex,
-                            charBuffer,
-                            0,
-                            numOfCharsToCopy);
-
-                        sourceIndex += numOfCharsToCopy;
+                        var charCount = Math.Min(Content.Length - sourceIndex, DefaultByteBufferSize);
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: I tried `resolvedContentTypeEncoding.GetMaxCharCount` but that resulted in exceptions in few cases (because the count of characters was greater than the buffer size)

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 4]
```diff
@@ -67,13 +68,30 @@ public override Task ExecuteResultAsync(ActionContext context)
 
             if (Content != null)
             {
-                var bufferingFeature = response.HttpContext.Features.Get<IHttpBufferingFeature>();
-                bufferingFeature?.DisableResponseBuffering();
+                response.ContentLength = resolvedContentTypeEncoding.GetByteCount(Content);
 
-                return response.WriteAsync(Content, resolvedContentTypeEncoding);
-            }
+                var byteBuffer = ArrayPool<byte>.Shared.Rent(DefaultByteBufferSize);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: I would suggest that you put the body of this action result into an action result executor facade (see ViewResult, FileResult, etc.) and require the array pool instance on the constructor of the executor rather than relying on the static instance. That's the pattern we've followed elsewhere.

/cc @rynowak

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 5]
```diff
@@ -20,6 +19,8 @@ public class ContentResult : ActionResult
             Encoding = Encoding.UTF8
         }.ToString();
 
+        private const int BufferSize = 1024;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: Misnamed. This is the number of characters to encode in a given pass.

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs [hunk 6]
```diff
@@ -67,13 +68,31 @@ public override Task ExecuteResultAsync(ActionContext context)
 
             if (Content != null)
             {
-                var bufferingFeature = response.HttpContext.Features.Get<IHttpBufferingFeature>();
-                bufferingFeature?.DisableResponseBuffering();
+                response.ContentLength = resolvedContentTypeEncoding.GetByteCount(Content);
 
-                return response.WriteAsync(Content, resolvedContentTypeEncoding);
-            }
+                var requiredLength = resolvedContentTypeEncoding.GetMaxByteCount(BufferSize);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: What if ContentLength is already known to be smaller than this?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: We're using the buffer pool here, so it seems unimportant

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs [hunk 7]
```diff
@@ -43,35 +42,6 @@ public class ContentResultTest
             MediaTypeAssert.Equal("text/plain; charset=utf-7", httpContext.Response.ContentType);
         }
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: We need some more tests here, particularly around the boundary conditions of the buffer size/chunk size.

File: src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs]: Agreed, this should have a better name.

**Comment by rynowak** [comment]
This comment from @javiercn got buried, but this is what we've been doing with other action results that need services:

>I would suggest that you put the body of this action result into an action result executor facade (see ViewResult, FileResult, etc.) and require the array pool instance on the constructor of the executor rather than relying on the static instance. That's the pattern we've followed elsewhere.

This is valuable because in general an action result needs to do a `GetRequiredService()` somewhere, but this limits it to a known composition root.

**Comment by kichalla** [comment]
ping

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        public const int MaxCharacterChunkSize = 1024;
+
+        private readonly string DefaultContentType = new MediaTypeHeaderValue("text/plain")
+        {
+            Encoding = Encoding.UTF8
+        }.ToString();
+
+        private readonly ILogger<ContentResultExecutor> _logger;
+
+        public ContentResultExecutor(ILogger<ContentResultExecutor> logger)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: The `ArrayPool` should also come from DI

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs [hunk 9]
```diff
@@ -167,6 +138,66 @@ public class ContentResultTest
...
+                    {  new string('a', maxCharacterChunkSize * 3), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize - 1), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize + 1), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize - 2), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize + 2), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize - 3), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize + 3), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize * 2), "text/plain; charset=utf-8" },
+                    {  new string('色', maxCharacterChunkSize * 3), "text/plain; charset=utf-8" }
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: It would also be good to see some 1-2 other encodings used here. It looks like you planned for that but didn't do it.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 8 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Sure, I noticed that we already register the following in DI..will fix
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L207-L208

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+            if (result == null)
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+
+            var response = context.HttpContext.Response;
+
+            string resolvedContentType;
+            Encoding resolvedContentTypeEncoding;
+            ResponseContentTypeHelper.ResolveContentTypeAndEncoding(
```
Comment (lodejard) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Looks great so far, but I don't like how this helper is implemented. 

1. it always reads the response.ContentType even when result.ContentType is set
1. it always assigns the response.ContentType even if the result.ContentType is not set
1. MediaType.GetEncoding is always called for the default, which is always UTF-8

I'd suggest changing the helper 
```cs
public static Encoding ApplyActionResultContentType(
  HttpResponse response, 
  string actionResultContentType = null, 
  Encoding defaultEncoding = null)
{
  if (!string.IsNullOrEmpty(actionResultContentType))
  {
    response.ContentType = actionResultContentType;
    return 
      MediaType.GetEncoding(actionResultContentType) ?? 
      defaultEncoding ??
      Encoding.UTF8;
  }

  var responseContentType = response.ContentType;
  if (!string.IsNullOrEmpty(responseContentType))   
  {
    return 
      MediaType.GetEncoding(responseContentType) ?? 
      defaultEncoding ?? 
      Encoding.UTF8;
  }

  return defaultEncoding ?? Encoding.UTF8;
}
```

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ContentResultExecutor
+    {
+        /// <summary>
+        /// The maximum number of characters that are encoded and written to stream in a single pass.
+        /// </summary>
+        public const int MaxCharacterChunkSize = 1024;
+
+        private readonly string DefaultContentType = new MediaTypeHeaderValue("text/plain")
```
Comment (lodejard) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Change to a string constant - we don't need to use a class to build this value, and certainly don't need to do it per-request.

Comment (lodejard) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Wait - I need to correct myself - based on my suggestion to change the helper method below this default can be removed entirely.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 10 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: > 1. it always reads the response.ContentType even when result.ContentType is set

There is no perf cost in reading response.ContentType, right? as its a string only

> 2. it always assigns the response.ContentType even if the result.ContentType is not set

If a user does `Content("Hello World")`, I would expect the response to have the default content type set (in this case `text/plain; charset=utf-8`, right?

Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Please, not perf arguments without measuring

**Comment by kichalla** [comment]
ping

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+                var byteBuffer = _byteArrayPool.Rent(requiredLength);
+
+                try
+                {
+                    var sourceIndex = 0;
+
+                    while (sourceIndex < result.Content.Length)
+                    {
+                        var charCount = Math.Min(result.Content.Length - sourceIndex, MaxCharacterChunkSize);
+
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: nit: too many whitelines

**Comment by javiercn** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Nothing here is per-request, this is a singleton service. I agree in general though, it would be more obvious to see the content type written out in text.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: I had a chance to catch up with @lodejard yesterday, his suggestion is centered around changing the behavior to not have a default content type. 

Please open a separate issue for that discussion, that's a much larger topic than what we're doing in this PR.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs (hunk 9 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: I'd also like to see a few more tests with multiples of the chunk size **and** a few characters +/-

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 10 above)
Comment (lodejard) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Yeah - 2 is functional - I was expecting a plain-old-string sans content-type, but it put text/plain for me, and there was no way to null that out.

Tbh we just returned a string in the end which became an object result, and that avoided the buffer disabling also.

Ah, that's an interesting question - now that the content result writing is all nice and shiny - is there a string output formatter that has the same technique in it?

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: > Ah, that's an interesting question - now that the content result writing is all nice and shiny - is there a string output formatter that has the same technique in it?

We'd also want to do this with response.WriteAsync(...): https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/HttpResponseWritingExtensions.cs#L63

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: So @lodejard - is what you're looking for here just a way to **not** set a content-type?


> Ah, that's an interesting question - now that the content result writing is all nice and shiny - is there a string output formatter that has the same technique in it?

String output formatter is a problematic IMO. I really don't like the way it works today and we've gotten lots of negative feedback about it. Sadly I think we missed our chance to get it right.

**Comment by kichalla** [comment]
:up: :date:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs [hunk 13]
```diff
@@ -167,6 +142,142 @@ public class ContentResultTest
             var finalResponseContentType = httpContext.Response.ContentType;
             Assert.Equal(expectedContentType, finalResponseContentType);
             Assert.Equal(expectedContentData, memoryStream.ToArray());
+            Assert.Equal(expectedContentData.Length, httpContext.Response.ContentLength);
+        }
+
+        public static TheoryData<string, string> ContentResult_WritesDataCorrectly_ForDifferentContentSizesData
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: @rynowak I feel these tests need to be ideally in HttpAbstractions(I checked the tests there and I think they are missing some coverage (ex: surrogate pair tests)) as they are more like testing the response stream writer. We could keep few sanity level tests here, thoughts?

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: If you found a gap, then we should add these tests to HttpAbstractions

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            _logger.ContentResultExecuting(resolvedContentType);
+
+            if (result.Content != null)
+            {
+                response.ContentLength = resolvedContentTypeEncoding.GetByteCount(result.Content);
+
+                using (var textWriter = _httpResponseStreamWriterFactory.CreateWriter(response.Body, resolvedContentTypeEncoding))
+                {
+                    await textWriter.WriteAsync(result.Content);
+                    await textWriter.FlushAsync();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: It might be worth a comment here to explain that flushing the `HttpResponseStreamWriter` does **not** flush the underlying stream. This just flushes the buffered text in the writer.

We do this rather than letting dispose handle it because dispose would call `Write` and we want to call `WriteAsync`

**Comment by rynowak** [comment]
:shipit: - but let's have a quick discussion about what tests go where.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs (hunk 14 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: @kichalla add that comment in the codez

Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs]: Sure, I agree

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs [hunk 15]
```diff
@@ -167,6 +142,142 @@ public class ContentResultTest
...
+
+        [Theory]
+        [InlineData(DefaultCharacterChunkSize)]
+        [InlineData(DefaultCharacterChunkSize * 2)]
+        [InlineData(DefaultCharacterChunkSize * 3)]
+        public async Task ContentResult_WritesDataCorrectly_ForCharactersHavingSurrogatePairs(int characterSize)
+        {
+            // Arrange
+            // Here "𐐀" (called Deseret Long I) actually represents 2 characters. Try to make this character split across
+            // the boundary
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ContentResultTest.cs]: :heart_eyes:
